Articulate EIP parameters and feature implementations

whilei opened the pull request:
This is essentially the same patch https://github.com/ethereum/go-ethereum/pull/18401.

Here's a copy-paste of the text from that PR's comment body:

---

Refactors chain configuration and respective feature implementations to use `IsEIP<NUMBER>` definitions and methods, instead of `Is<HardForkName>`, whenever possible. Doing so attempts to address ambiguity and complexity in chain configuration and feature implementation.

As I see it, the __benefits are__:

1. More descriptive code. By describing and implementing client configuration with feature-based definitions, instead of only arbitrary and opaque hard-fork feature groups, feature implementations become clearer. This improves the code's legibility and accessibility, separates logical concerns, documents specification references, and allows more granular testing.

2. It's more interoperable. Clients choosing to adopt a subset of EIP-derived changes, normally inextricably bundled in a hard-fork identity, are able to toggle individual features. This establishes an extensible pattern that alternative implementations can use to build clients with supersets or subsets of features. 

3. It doesn't break backwards compatibility. All existing hardcoded or external chain configurations continue to operate as expected, tests pass, and named hard-fork keys and methods take priority.

What this patch __doesn't do__: 

1. Handle every feature ever introduced via the EIP/Hard-fork processes. For example, of the changes introduced with the `Homestead` fork, only EIP 7's `DELEGATECALL` has been extracted, and the DAO fork is untouched. Just low-hanging fruit.

2. Implement correlated refactoring across the `tests/[testdata/**/*.json]` tests and test runners. These are located in a submodule, very numerous, very opinionated toward the hardfork schema, and their relevance and applicability is not impacted by leaving them as-is.

3. Implement new distinct difficulty calculators for [EIP100 (Change difficulty adjustment to target mean block time including uncles)](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-100.md) vs. [EIP649 (Delaying the difficulty bomb and reducing the block reward)](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-649.md), which were two modifications to the difficulty algorithm introduced simultaneously at the [Byzantium hard-fork](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-609.md). I'm trying to limit the scope of changes to introduce as little new consensus logic as possible.

4. Attempt to differentiate features beyond the specifications of EIPs. 

5. Move beyond supplemental modification of existing configuration patterns.

---

For review and reference, I compiled [this gist for Hardfork/EIP reference](https://gist.github.com/whilei/e86d7377df636df1a01ac316f34961a0) along the way.

---

whilei committed:
cmd,consensus,core,eth,params,tests: articulate EIP features

* NOTE: This is a cherry-picked commit, and has undergone
some manual merge conflict resolution. It may not be perfectly
perfect yet.

The original commit body:

Refactors chain configuration and respective feature
implementations to use EIP definitions and methods,
instead of HardFork names, whenever possible. Doing so
attempts to address ambiguity and complexity in chain
configuration and feature implementation.

Signed-off-by: Isaac Ardis (isaac.ardis@gmail.com)

consensus: fix wrong numbers for blockreward eips

IDK how these got off-by-one'd. Weird.

params: alphabetize ecip blocks

Because I'm anal like that.

consensus/ethash: fix mysterious off-by-one reward

Same as where the EIP number got incremented down. Weird.

params: note about json tag for eip160(F)Block

This is a bit of a sticky situation. Solves are either:
- use without 'F' here, break pattern, stay backwards
compatible for existing multi-geth clients, but incompatible
with ETH if they accept proposed changes as-is
- use with 'F' here, break backwards compatibility with
existing multi-geth clients, but be compatible with ETH as proposed
- push change without 'F' to ETH for review, keeping
in mind that it makes no sense for them to tangle their
code b/c of an establish multi-geth pattern
- push change without 'F' for any of the EIP160,161,170 to
eth. still breaks pattern for them

The root cause of this is that EIP158 namespace was used
to symbolize 3 separate EIPs... none of which ended up being
named 'EIP158'. Yea.

params: fix comment to be right

consensus/ethash: fix another damn offbyone increment

... these happened during the cherry-pick. wtf.

whilei commented on params/config.go:
```diff
@@ -255,30 +399,102 @@ type TrustedCheckpoint struct {
+	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-160.md
+	// NOTE: this json tag (a.) varies from it's 'siblings', which have 'F's in them, and (b.) without the 'F' will vary from ETH implementations if they choose to accept the PR
+	// FIXME?
+	EIP160FBlock *big.Int `json:"eip160Block,omitempty"`
```
@sorpaas PTAL  This json tag business is a bit of a sticky situation.

The root cause of this is that the EIP158 namespace was used to symbolize a hardfork, including 3 separate EIPs (erm... none of which ended being named EIP158, to futher complicate the matter). 
So articulating params/features by an `EIPxxx` namespace gets messy. This was the whole reason I moved to using the `F` in the first place, where "F" stands for Feature.

Possible solves include...
- use without 'F' here, break json naming pattern, stay backwards
compatible for existing multi-geth clients, but incompatible
with ETH if they accept proposed changes as-is
- use with 'F' here, break backwards compatibility with
existing multi-geth clients, but be compatible with ETH as proposed
- push change without 'F' to ETH for review, keeping
in mind that it makes no sense for them to tangle their
code b/c of an establish multi-geth pattern
- push change without 'F' for any of the EIP160,161,170 to
eth. still breaks json naming pattern
- any other you can think of?...

sorpaas replied:
I have a slight preference that we use without `F` (the first choice), **and** remove all grouped hard fork parameters (`ConstantinopleBlock`, `ByzantiumBlock`, etc).

* 1.8.20 will be the last version in the 1.8 series, meaning we can do breaking changes.
* Keeping both grouped features (`ConstantinopleBlock` etc) and individual features together (`EIP1234Block` etc) add maintenance burden, because we always need to check two booleans instead of one.
* For config parameters, keeping it in line with upstream may not be that important. The issue is that every time when config parameter changes, we need to resolve merge conflicts manually anyway.

But it's up to you!

whilei replied:
Roger that. 

Agreed on the `F` choice; I'll leave that as-is then. 

I'm looking at the implications of `IsConstantinople`/`ConstantinopleBlock` changes now. I definitely understand the motivations for removing them, and the appeal of a through-and-through consistent configuration pattern is enticing. My biggest hesitation is that using "groups" gives a way to "alias" often-reused configuration bunches, and removing them will require adding quite a few lines of repeated configurations which is both harder to read and, without adequate comments, possibly harder to understand....  I'm going to do a little exploring to see what the implications would look like.

whilei reviewed (commented):
I never know if this part is necessary for my inline comment below. I want to click `Request changes` since I'm waiting for review... but only letting me click `Comment`. 

Anyways, just awaiting feedback on the comment.

whilei committed:
core/vm: gofmt -w -s contracts_test.go b/c travis's linter

whilei committed:
params: Edit comment to reflect decision about 'F' name scheme

whilei committed:
params: EIP2F: add configuration parameters

This is the remaining not-backwards-compatible Homestead
EIP not yet 'feature-fied'.

The EIP itself specifies a diverse set of changes, those
will be configuration-implemented in the next commit.

whilei committed:
consensus,core/light: EIP2F: implement EIP2F instead of Homestead features implementations

whilei committed:
params: update comment to reflect eip2 is now feature-configured

sorpaas commented on consensus/ethash/consensus.go:
```diff
@@ -320,13 +322,19 @@ func CalcDifficulty(config *params.ChainConfig, time uint64, parent *types.Heade
+	case config.IsByzantium(next) || (config.IsEIP649F(next) && config.IsEIP100F(next)):
+		return calcDifficultyByzantium(time, parent)
+	case config.IsEIP649F(next):
+		// TODO: calculator for only EIP649:difficulty bomb delay (without EIP100:mean time adjustment)
```
```suggestion
		// TODO (#22): calculator for only EIP649:difficulty bomb delay (without EIP100:mean time adjustment)
```

sorpaas commented on consensus/ethash/consensus.go:
```diff
@@ -320,13 +322,19 @@ func CalcDifficulty(config *params.ChainConfig, time uint64, parent *types.Heade
+		// TODO: calculator for only EIP649:difficulty bomb delay (without EIP100:mean time adjustment)
+		panic("not implemented")
+	case config.IsEIP100F(next):
+		// TODO: calculator for only EIP100:mean time adjustment (without EIP649:difficulty bomb delay)
```
```suggestion
		// TODO (#23): calculator for only EIP100:mean time adjustment (without EIP649:difficulty bomb delay)
```

sorpaas commented on core/vm/interpreter.go:
```diff
@@ -131,6 +123,36 @@ func (in *EVMInterpreter) enforceRestrictions(op OpCode, operation operation, st
 			}
 		}
 	}
+	switch op {
```
What's the reason we switch from using `newInstructionSet` for restrictions?

whilei replied:
`newInstructionSet` now includes ALL available opcodes, where before there was a dedicated instruction set for each hardfork. The logic before checked for the nonexistance of a given opcode, and [this line](https://github.com/ethoxy/multi-geth/pull/19/files/be858d0b67dfb139e669e7c2512abf3cd1e10ff3#diff-8f9edb073f33b63dc5736262579687b8R238) would fail on a "nonexistant" code. 

Continuing this pattern would have been impossible because it presents a permutation problem when arbitrary opcode sets can be configured.

sorpaas replied:
> Continuing this pattern would have been impossible because it presents a permutation problem when arbitrary opcode sets can be configured.

Does something like this work?

```
instructionSet := newBasicInstructionSet()
if config.IsEIP1052() {
  instructionSet[EXTCODEHASH] = ...
}
```

whilei replied:
Yea that would work. One of my initial sketches used this pattern, but I moved away from it thinking the current way seemed 'simpler', with a single resource for known opcodes that I thought might be useful for testing, printing, ... but nothing definite.

Happy to move to this pattern, though. Are you thinking it'd be faster since it avoids the `case` switch for each op (and shifts the expense to a single overhead per interpreter)?

sorpaas replied:
It may have slight performance difference, but I just think that using the above pattern might be more future-proof -- what if by accident, we define slightly different opcode on different networks with the same bytecode. WDYT?

whilei replied:
Ah, different opcode(s) behavior across networks with same bytecode... I hadn't considered that. Alright, will refactor.

whilei replied:
See 257cf96

sorpaas committed:
consensus/ethash/consensus.go: add issue reference

Co-Authored-By: whilei <isaac.ardis@gmail.com>

sorpaas committed:
consensus/ethash/consensus.go: add issue reference

Co-Authored-By: whilei <isaac.ardis@gmail.com>

whilei committed:
core/vm: refactor instruction set to build per chain config

This allows for the potential for opcode mechanisms to
be defined differently even if they share the same bytecode.

sorpaas commented on core/vm/contracts.go:
```diff
@@ -59,6 +49,27 @@ var PrecompiledContractsByzantium = map[common.Address]PrecompiledContract{
 }
 
+// IsPrecompiledContractEnabled checks whether a given precompiled contract is enabled for a chain config at a given block.
+func IsPrecompiledContractEnabled(config *params.ChainConfig, num *big.Int, codeAddr common.Address) bool {
```
Sorry I missed this, but do you mind if we do the same for precompiled contracts?

whilei replied:
Ah, yeap. Will do.

whilei replied:
Pending review, resolved w/ da85a44.

sorpaas reviewed (commented):
I suggest we change precompiled contract enable-ment the same we we did for instruction set. All rest looks good to me!

whilei committed:
core/vm,eth/tracers: use per config precompiled map

This is akin to 257cf96, allowing variability in precompiled
contract mechanisms across network configurations.
